GiftCertificatesSystem/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,112 @@
+package com.epam.esm.service.impl;
+
+import com.epam.esm.dao.GiftCertificateDao;
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.dao.GiftCertificateQuery;
+import com.epam.esm.service.GiftCertificateService;
+import com.epam.esm.service.exception.GiftCertificateNotFoundServiceException;
+import com.epam.esm.service.exception.InvalidRequestedIdServiceException;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+    private final GiftCertificateDao giftCertificateDao;
+
+    public GiftCertificateServiceImpl(GiftCertificateDao giftCertificateDao) {
+        this.giftCertificateDao = giftCertificateDao;
+    }
+
+    @Override
+    @Transactional
+    public GiftCertificate addGiftCertificate(GiftCertificate giftCertificate) {
+        LocalDateTime localDateTime = LocalDateTime.now();
+        giftCertificate.setCreateDate(localDateTime);
+        giftCertificate.setLastUpdateDate(localDateTime);
+        giftCertificateDao.save(giftCertificate);
+        addGiftCertificateTags(giftCertificate);
+        return giftCertificate;
+    }
+
+    @Override
+    public GiftCertificate getGiftCertificateById(Long id)
+            throws GiftCertificateNotFoundServiceException, InvalidRequestedIdServiceException {
+        if (id > 0) {
+            Optional<GiftCertificate> giftCertificateById = giftCertificateDao.findById(id);
+            return giftCertificateById.orElseThrow(() ->
+                    new GiftCertificateNotFoundServiceException("GiftCertificate with id=" + id + " not found")
+            );
+        } else {
+            throw new InvalidRequestedIdServiceException(id + " does not fit the allowed gap. Expected gap: 0 > id");
+        }
```

Comment:
```suggestion
        if (id <= 0) {
            throw new InvalidRequestedIdServiceException(id + " does not fit the allowed gap. Expected gap: 0 > id");
        }
         
        Optional<GiftCertificate> giftCertificateById = giftCertificateDao.findById(id);
        return giftCertificateById.orElseThrow(() ->
                    new GiftCertificateNotFoundServiceException("GiftCertificate with id=" + id + " not found")
            );
```

Please consider validation method to check id.